### ISSUE TITLE
Unable to Create @Page Object using Spring Boot 2

### ISSUE THREAD

**Issue by cjaehnen** [FluentLenium/FluentLenium#584]
I'm currently converting a Spring Boot app from 1.5.x to 2.0.0. When I run my Fluentlenium test, I get the below exception.

Major differences include I'm using JUnit 5 instead of 4. Any ideas?

```
java.lang.NullPointerException
	at com.gaig.cloudnative.test.app.acceptance.base.AppFluentTest.setUp(AppFluentTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:436)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.invokeMethodInExtensionContext(ClassTestDescriptor.java:302)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$synthesizeBeforeEachMethodAdapter$12(ClassTestDescriptor.java:290)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachMethods$2(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:155)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachMethods(TestMethodTestDescriptor.java:134)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:109)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:112)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Here's my code:

AppFluentTest.java

```java
import com.gaig.cloudnative.test.app.TestApplication;
import com.gaig.cloudnative.test.app.acceptance.page.HomePage;
import com.google.common.io.PatternFilenameFilter;
import lombok.extern.slf4j.Slf4j;
import org.fluentlenium.adapter.junit.FluentTest;
import org.fluentlenium.configuration.FluentConfiguration;
import org.fluentlenium.core.annotation.Page;
import org.fluentlenium.core.domain.FluentWebElement;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeDriverService;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.net.URL;
import java.util.*;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Fail.fail;
import static org.fluentlenium.configuration.ConfigurationProperties.DriverLifecycle.CLASS;
import static org.fluentlenium.configuration.ConfigurationProperties.TriggerMode.AUTOMATIC_ON_FAIL;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
import static org.springframework.test.annotation.DirtiesContext.ClassMode.AFTER_CLASS;
import static org.springframework.test.context.jdbc.Sql.ExecutionPhase.AFTER_TEST_METHOD;
import static org.springframework.test.context.jdbc.Sql.ExecutionPhase.BEFORE_TEST_METHOD;

@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = RANDOM_PORT, classes = TestApplication.class)
@ActiveProfiles("h2")
@DirtiesContext(classMode = AFTER_CLASS)
@Sql(executionPhase = BEFORE_TEST_METHOD, scripts = "classpath:db/h2/reset.sql")
@Sql(executionPhase = AFTER_TEST_METHOD, scripts = "classpath:db/h2/reset.sql")
@FluentConfiguration(driverLifecycle = CLASS, screenshotMode = AUTOMATIC_ON_FAIL, screenshotPath = "screenshots")
@Slf4j
public abstract class AppFluentTest extends FluentTest {

    private final List<File> filesToDelete = new ArrayList<>();

    @LocalServerPort
    protected int serverPort;

    @Page
    protected HomePage homePage;

    @BeforeEach
    public void setUp() {
        String baseUrl = getBaseUrl(serverPort);
        homePage.setBaseUrl(baseUrl);
    }

    @AfterEach
    public void tearDown() {
        filesToDelete.forEach(File::delete);
    }

    @Override
    public WebDriver newWebDriver() {
        try {
            return new ChromeDriver(getChromeDriverService(), getChromeOptions());
        } catch (Exception e) {
            return null;
        }
    }

    protected WebDriverWait waitForChromeDriver(int delayMilliseconds) {
        return new WebDriverWait(getDriver(), delayMilliseconds, 10);
    }

    protected String getBaseUrl(int serverPort) {
        log.debug("Base url using server port {}", serverPort);
        return String.format("http://%s:%s@localhost:%s", "user@gaig.com", "password", serverPort);
    }

    protected String getCurrentUrl() {
        return getDriver().getCurrentUrl();
    }

    protected void assertUrlContains(String url) throws Exception {
        assertThat(new URL(getCurrentUrl()).getPath()).contains(url);
    }

    protected void assertUrl(String expectedUrl) throws Exception {
        assertUrl(getCurrentUrl(), expectedUrl);
    }

    protected void assertUrl(String actualUrl, String expectedUrl) {
        assertThat(actualUrl).isEqualTo(expectedUrl);
    }

    protected File getTmpFile(String startsWith) {
        return getTmpFile(new PatternFilenameFilter(String.format("%s.*[^crdownload]", startsWith)));
    }

    protected File getTmpFile(FilenameFilter filter) {
        int attempts = 0;
        while (attempts < 20) {
            try {
                File directory = new File(new File(".").getCanonicalPath() + "/tmp");
                if (directory.listFiles() != null) {
                    File[] files = directory.listFiles(filter);
                    if (files != null && files.length > 0) {
                        if (files.length > 1) {
                            log.debug("More than one file found. Returning index 0.");
                        }
                        log.debug("Returning file {}", files[0].getName());
                        filesToDelete.add(files[0]);
                        return files[0];
                    }
                }
                Thread.sleep(200);
            } catch (IOException e) {
                fail("Could not get file");
            } catch (InterruptedException e) {
                fail("Something went wrong fetching the file");
            }
            attempts++;
        }
        return null;
    }

    protected void switchToTab(int tab) {
        WebDriver driver = getDriver();

        for (int i = 2; i >= 0; i--) {
            ArrayList<String> tabs = new ArrayList<>(getDriver().getWindowHandles());
            if (tabs.size() > tab) {
                driver.switchTo().window(tabs.get(tab));
                return;
            }
            try {
                log.debug("Failed to switch tabs. Sleeping 100 milliseconds.");
                Thread.sleep(100);
            } catch (InterruptedException ignored) {
            }
        }
        throw new RuntimeException("Could not switch to tab #" + tab);
    }

    protected void closeTab(int tab) {
        WebDriver driver = getDriver();

        for (int i = 2; i >= 0; i--) {
            ArrayList<String> tabs = new ArrayList<>(getDriver().getWindowHandles());
            if (tabs.size() > tab) {
                driver.switchTo().window(tabs.get(tab)).close();
                switchToTab(tab - 1);
                return;
            }
        }
        throw new RuntimeException("Could not switch to tab #" + tab);
    }

    protected void clearDatePicker(FluentWebElement element) {
        element.keyboard().sendKeys(Keys.chord(
            Keys.BACK_SPACE,
            Keys.LEFT,
            Keys.BACK_SPACE,
            Keys.LEFT,
            Keys.BACK_SPACE));
    }

    private ChromeDriverService getChromeDriverService() {
        return new ChromeDriverService.Builder()
            .usingAnyFreePort()
            .build();
    }

    private ChromeOptions getChromeOptions() throws Exception {
        Map<String, Object> chromePreferences = new HashMap<>();
        chromePreferences.put("profile.default_content_settings.popups", 0);
        chromePreferences.put("download.default_directory", new File(".").getCanonicalPath() + "/tmp");

        DesiredCapabilities capabilities = DesiredCapabilities.chrome();
        capabilities.setCapability(CapabilityType.ACCEPT_SSL_CERTS, true);

        ChromeOptions options = new ChromeOptions();
        options.setExperimentalOption("prefs", chromePreferences);
        options.addArguments("--window-size=3000,1440");

        if (isCI()) {
            options.addArguments(asList("headless", "disable-gpu", "no-sandbox"));
        }

        System.setProperty("webdriver.chrome.silentOutput", "true");

        options.merge(capabilities);
        return options;
    }

    private boolean isCI() {
        return Optional.ofNullable(System.getenv("CI")).orElse("false").equals("true");
    }
}
```

AppPage.java

```java
import lombok.extern.slf4j.Slf4j;
import org.fluentlenium.core.FluentControl;
import org.fluentlenium.core.FluentPage;
import org.fluentlenium.core.domain.FluentList;
import org.fluentlenium.core.domain.FluentWebElement;
import org.openqa.selenium.By;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.function.Function;

@Slf4j
public abstract class AppPage extends FluentPage {

    @Override
    public void initFluent(FluentControl control) {
        super.initFluent(control);
    }

    public WebDriverWait waitForChromeDriver(int delaySeconds) {
        return new WebDriverWait(getDriver(), delaySeconds, 200);
    }

    public void refreshPage() {
        getDriver().navigate().refresh();
    }

    public void expandFirstAccordion(String selector) {
        FluentList<FluentWebElement> accordions = find("li" + selector);
        FluentWebElement element = accordions.first();
        clickAndWait(element);
    }

    public void expandAllAccordions() {
        expandAllAccordions(".accordion-item");
    }

    public void expandAllAccordions(String selector) {
        FluentList<FluentWebElement> accordions = find(selector);
        for (FluentWebElement element : accordions) {
            if (!element.attribute("class").contains("is-active")) {
                clickAndWait(element);
            }
        }
    }

    public void expandAllChildAccordions(String parentSelector) {
        expandAllAccordions(parentSelector + " .accordion-item");
    }

    private void clickAndWait(FluentWebElement element) {
        element.scrollIntoView();

        try {
            Thread.sleep(90);
        } catch (InterruptedException ignored) {
        }

        if (element.clickable()) {
            try {
                element.click();
            } catch (WebDriverException e) {
                log.debug("Exception occurred while clicking");
            }

            try {
                waitForChromeDriver(5).until((Function) driver ->
                    ExpectedConditions.presenceOfNestedElementLocatedBy(
                        element.getElement(),
                        By.cssSelector(".accordion-content > *")));
            } catch (TimeoutException e) {
                log.debug("Timeout occurred waiting for table wrapper to appear");
            }

            try {
                Thread.sleep(20);
            } catch (InterruptedException ignored) {
            }
        }
    }
}
```

HomePage.java

```java
import org.fluentlenium.core.annotation.PageUrl;
import org.fluentlenium.core.domain.FluentWebElement;
import org.openqa.selenium.support.FindBy;

@PageUrl("/secure/home")
public class HomePage extends AppPage {

    @FindBy(css = "h2")
    private FluentWebElement title;

    public String getTitle() {
        return title.text();
    }
}
```

HomePageTest.java

```java
import com.gaig.cloudnative.test.app.acceptance.base.AppFluentTest;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class HomePageTest extends AppFluentTest {

    @Test
    void showsHomePage() throws Exception {
        goTo(homePage);

        assertThat(homePage.getTitle()).isEqualTo("Home");
    }
}
```

**Comment by filipcynarski** [comment]
Have you tried to run your tests using JUnit4? Currently we are working on JUnit5 module in next release. I think it may be a case. I'll try to run your code later.

**Comment by cjaehnen** [comment]
I believe you are correct, as I ran the test in JUnit 4, and got a different error but the Page class is no longer null. I'll have to revert to JUnit 4 for these types of tests, for now.

Any idea when you expect JUnit 5 support?

Thanks for your assistance. I love FluentLenium!

**Comment by filipcynarski** [comment]
I'm glad to hear that! The work is in progress https://github.com/FluentLenium/FluentLenium/pull/556 we are planning to release new version in next 2 months.

**Comment by filipcynarski** [comment]
@cjaehnen would you be able is your issue resolved? I've added JUnit Jupiter in 3.7.1 release

**Comment by slawekradzyminski** [comment]
Closing - no response
